src/lib.rs

```diff
@@ -144,79 +125,27 @@ pub struct Options<'a, S = Box<dyn WordSplitter>> {
     pub splitter: S,
 }
 
-/// Allows using an `Options` with [`wrap`] and [`fill`]:
-///
-/// ```
-/// use textwrap::{fill, Options};
-///
-/// let options = Options::new(15).initial_indent("> ");
-/// assert_eq!(fill("Wrapping with options!", &options),
-///            "> Wrapping with\noptions!");
-/// ```
-///
-/// The integer specifes the wrapping width. This is equivalent to
-/// passing `Options::new(15)`.
-///
-/// [`wrap`]: fn.wrap.html
-/// [`fill`]: fn.fill.html
-impl<S: WordSplitter> WrapOptions for &Options<'_, S> {
-    #[inline]
-    fn width(&self) -> usize {
-        self.width
-    }
-    #[inline]
-    fn initial_indent(&self) -> &str {
-        self.initial_indent
-    }
-    #[inline]
-    fn subsequent_indent(&self) -> &str {
-        self.subsequent_indent
-    }
-    #[inline]
-    fn break_words(&self) -> bool {
-        self.break_words
-    }
-    #[inline]
-    fn split_points(&self, word: &str) -> Vec<usize> {
-        self.splitter.split_points(word)
+impl<'a, S> From<&'a Options<'a, S>> for Options<'a, &'a S> {
```

Comment:
This looks a bit weird to me — it's basically saying that you can go from `&Options` to `Options` easily. This is the same as what `*options` would do, right? So would we avoid it somehow?

Reply:
It's different because it also takes a reference to `S` which removes the need for it to be `Clone`. I also wanted to support passing `&Config` to the functions as it avoids breaking `split_words` too much, which used to take an `&impl WrapOptions`.

Reply:
Ah, I hadn't spotted the `S` to `&'a S` change :-)

As for breaking `split_words` — feel free to break it all you want, it's a brand new function and I'm happy to hammer out a better signature before I release it to the wild.

Reply:
`wrap` internally calls `split_words` multiple times. If this was not provided then it would have to (a) clone the `Config` each time and bound `S: Clone` or (b) manually write out the body of this impl, both of which this impl seems preferable to.

That being said, I agree that this impl it is a bit weird. If ergonomics was not an issue all the functions would take `&Options<'_, S>`. However, we want to support `usize`, `Options<'_, S>` _and_ `&Options<'_, S>`. But it's hard to find anything that they all have in common:
* `usize` supports conversion from a reference or value to owned options.
* `Options<'_, S>` supports conversion from a reference to shared options or a value to owned options, or a reference to owned options if we include conversions to `Options<'_, &S>`.
* `&Options<'_, S>` supports conversion from a reference or value to shared options, or to owned options if we include conversions to `Options<'_, &S>`.

The current solution converts a value to owned options, as that's the only thing supported by all three. But you do end up with weird double-references sometimes.

Reply:
Thanks for that overview, makes this much clearer! It seems I've managed to come up with a bit of a "interesting" API with my weird insistence that `wrap` and `fill` can take both `usize` and some-form-of-options (don't care much here) :smile: 

As for double references... I'm not so concerned with them as long as they're internal to `textwrap` — in particular functions in `textwrap::core` which I consider a place for advanced functionality.